py/persipubsub/queue.py

```diff
@@ -0,0 +1,489 @@
+                    key=persipubsub.str_to_bytes(sub), txn=txn, create=False)
+                txn.put(key=persipubsub.str_to_bytes(msg_id), db=sub_db)
+
+    def put_many_flush_once(self, msgs: List[bytes]) -> None:
+        """
+        Put multiple message to LMDB in one transaction.
+
+        :param msgs: messages in bytes
+        :return:
+        """
+        # every publisher always prunes queue before sending a message.
+        self.vacuum()
+        assert isinstance(self.env, lmdb.Environment)
+        assert isinstance(self.subscriber_ids, List)
+        with self.env.begin(write=True) as txn:
+
+            pending_db = self.env.open_db(
+                key=persipubsub.PENDING_DB, txn=txn, create=False)
+
+            meta_db = self.env.open_db(
+                key=persipubsub.META_DB, txn=txn, create=False)
+
+            data_db = self.env.open_db(
+                key=persipubsub.DATA_DB, txn=txn, create=False)
+
+            sub_dbs = set()
+            for sub in self.subscriber_ids:
+                sub_dbs.add(
+                    self.env.open_db(
+                        key=persipubsub.str_to_bytes(sub),
+                        txn=txn,
+                        create=False))
+
+            for msg in msgs:
+                msg_id = str(datetime.datetime.utcnow().timestamp()) + str(
+                    uuid.uuid4())
+
+                txn.put(
+                    key=persipubsub.str_to_bytes(msg_id),
+                    value=persipubsub.int_to_bytes(len(self.subscriber_ids)),
+                    db=pending_db)
+
+                txn.put(
+                    key=persipubsub.str_to_bytes(msg_id),
+                    value=persipubsub.int_to_bytes(
+                        int(datetime.datetime.utcnow().timestamp())),
+                    db=meta_db)
+
+                txn.put(
+                    key=persipubsub.str_to_bytes(msg_id), value=msg, db=data_db)
+
+                for sub_db in sub_dbs:
+                    txn.put(key=persipubsub.str_to_bytes(msg_id), db=sub_db)
+
+    def front(self, identifier: str) -> Optional[bytes]:
+        """
+        Peek at next message in LMDB.
+
+        Load from LMDB into memory and process msg afterwards.
+
+        :param identifier: Subscriber ID
+        :return:
+        """
+        assert isinstance(self.env, lmdb.Environment)
+        with self.env.begin(write=False) as txn:
+            sub_db = self.env.open_db(
+                key=persipubsub.str_to_bytes(identifier), txn=txn, create=False)
+            data_db = self.env.open_db(
+                key=persipubsub.DATA_DB, txn=txn, create=False)
+
+            cursor = txn.cursor(db=sub_db)
+            # check if database is not empty
+            if cursor.first():
+                key = cursor.key()
+                msg = txn.get(key=key, db=data_db)
+            else:
+                msg = None
+
+        return msg  # type: ignore
+
+    def pop(self, identifier: str) -> None:
```

Comment:
Since the class is called _Queue, I'd rename the argument to `sub_id` to be consistent with other methods.